platforms/nuttx/src/px4/common/px4_crypto.cpp

```diff
@@ -0,0 +1,139 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 Technology Innovation Institute. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#if defined(PX4_CRYPTO)
+
+#include <px4_platform_common/crypto.h>
+#include <px4_platform_common/crypto_backend.h>
+
+extern "C" {
+#include <nuttx/random.h>
+}
+
+
+px4_sem_t PX4Crypto::_lock;
+bool PX4Crypto::_initialized = false;
+
+void PX4Crypto::px4_crypto_init()
+{
+	if (PX4Crypto::_initialized) {
+		return;
+	}
+
+	px4_sem_init(&PX4Crypto::_lock, 0, 1);
+
+	// Initialize nuttx random pool, if it is being used by crypto
+#ifdef CONFIG_CRYPTO_RANDOM_POOL
+	up_randompool_initialize();
+#endif
+
+	// initialize keystore functionality
+	keystore_init();
+
+	// initialize actual crypto algoritms
+	crypto_init();
+
+	PX4Crypto::_initialized = true;
+}
+
+PX4Crypto *PX4Crypto::px4_crypto_open(px4_crypto_algorithm_t algorithm)
```

Comment:
Apart from `px4_crypto_init`, I guess all of these are the same on all OSes? I'd move it to `platforms/common` then, so the OS only has to define the init.